File: scripts/deck_browser.mjs
// Actual isolated Mac Chrome checks; the Python controller registers success.
import fs from 'node:fs/promises';
import path from 'node:path';
import {pathToFileURL} from 'node:url';
import {createRequire} from 'node:module';
import {createHash} from 'node:crypto';

const require = createRequire(import.meta.url);
const modulePath = process.env.HARNESS_NODE_MODULES;
const [rootArg, requestArg] = process.argv.slice(2);
const root = path.resolve(rootArg);
function local(relative) {
  const resolved = path.resolve(root, relative);
  if (!relative || path.isAbsolute(relative) || !resolved.startsWith(root + path.sep))
    throw new Error('Project-relative path required');
  return resolved;
}
const request = JSON.parse(await fs.readFile(local(requestArg), 'utf8'));
const folder = path.dirname(local(requestArg));
const output = path.join(folder, 'browser-result.json');
const evidence = [], steps = [], errors = [], external = [];
const checks = {display:false, navigation:false, editing:false, export_reopen:false, offline:false};
let chromium;
let playwrightVersion;
async function ref(file) {
  return {path:path.relative(root, file).split(path.sep).join('/'),
    sha256:createHash('sha256').update(await fs.readFile(file)).digest('hex')};
}
function done(name, observed) {
  checks[name] = true;
  steps.push({name, passed:true, observed});
}
function watch(page) {
  page.on('pageerror', error => errors.push(error.message));
  // Save only origin; external URLs may contain credentials or private query data.
  page.on('request', req => {if (/^https?:/.test(req.url())) external.push(new URL(req.url()).origin);});
}
let browser;
try {
  if (process.platform !== 'darwin' || request.protocol !== 'harness-deck-browser-v1')
    throw new Error('Unsupported browser probe environment or protocol');
  const playwright = require(modulePath ? path.join(modulePath, 'playwright') : 'playwright');
  chromium = playwright.chromium;
  playwrightVersion = require(modulePath ? path.join(modulePath, 'playwright/package.json') :
    'playwright/package.json').version;
  const source = local(request.input_html.path);
  if ((await ref(source)).sha256 !== request.input_html.sha256) throw new Error('HTML fingerprint changed');
  browser = await chromium.launch({channel:'chrome', headless:true, chromiumSandbox:true});
  const environment = {platform:process.platform, browser_version:browser.version(),
    playwright_version:playwrightVersion, node:process.version};
  const context = await browser.newContext({offline:true, viewport:{width:1280,height:900}, acceptDownloads:true});
  const page = await context.newPage();
  watch(page);
  await page.goto(pathToFileURL(source).href);
  const data = JSON.parse(await page.locator('[data-phase5-deck]').textContent());
  if (JSON.stringify(data.deck_ref) !== JSON.stringify(request.target)) throw new Error('Wrong deck version');
  const count = data.pages.length;
  if (await page.locator('.slide').count() !== count) throw new Error('Slide count mismatch');
  for (const viewport of [{width:1280,height:900}, {width:390,height:844}]) {
    await page.setViewportSize(viewport);
    for (let i=0;i<count;i++) {
      await page.locator('#pages').selectOption(String(i));
      const active = page.locator(`#page-${i+1}`);
      if (!await active.isVisible() || await page.locator('.slide:visible').count() !== 1)
        throw new Error('Navigation visibility mismatch');
      const layout = await active.evaluate(slide => {
        const outer = slide.getBoundingClientRect();
        const fields = [...slide.querySelectorAll('[data-edit],.folio')]
          .filter(field => field.textContent.trim());
        return fields.map(field => {
          const box = field.getBoundingClientRect(), style = getComputedStyle(field);
          const range = document.createRange();
          range.selectNodeContents(field);
          const textRects = [...range.getClientRects()];
          const transparent = color => /rgba\([^)]*,\s*0(?:\.0+)?\s*\)$/.test(color);
          const hidden = !field.checkVisibility({checkOpacity:true,checkVisibilityCSS:true}) ||
            box.width <= 0 || box.height <= 0 || parseFloat(style.fontSize) < 1 ||
            transparent(style.color) || transparent(style.webkitTextFillColor);
          const clipped = textRects.some(r => r.left < box.left-1 || r.right > box.right+1 ||
            r.top < box.top-1 || r.bottom > box.bottom+1);
          const overlap = fields.some(other => {
            if (other === field) return false;
            const b = other.getBoundingClientRect();
            return Math.min(box.right,b.right)-Math.max(box.left,b.left)>1 &&
              Math.min(box.bottom,b.bottom)-Math.max(box.top,b.top)>1;
          });
          return {text:field.textContent, hidden, clipped, overlap,
            overflow:field.scrollWidth>field.clientWidth+1 || field.scrollHeight>field.clientHeight+1,
            outside:box.left<outer.left-1 || box.right>outer.right+1 || box.top<outer.top-1 || box.bottom>outer.bottom+1,
            belowViewport:box.bottom>window.innerHeight+1};
        });
      });
      // Narrow long pages can scroll; horizontal overflow and escaped fields fail.
      if (layout.some(x => x.hidden || x.clipped || x.overlap))
        throw new Error(`Hidden, clipped or overlapping content on page ${i+1}`);
      if (layout.some(x => x.overflow || x.outside)) throw new Error(`Overflow on page ${i+1}`);
      const text = await active.locator('[data-edit]').allTextContents();
      if (text[0] !== data.pages[i].title || text.slice(1).join('') !== data.pages[i].text)
        throw new Error(`Changed page-plan text on page ${i+1}`);
      const images = await active.locator('img').evaluateAll(items => items.map(img => ({
        loaded:img.complete && img.naturalWidth>0 && img.naturalHeight>0,
        visible:img.checkVisibility({checkOpacity:true,checkVisibilityCSS:true}),
        width:img.getBoundingClientRect().width,height:img.getBoundingClientRect().height
      })));
      if (images.some(img => !img.loaded || !img.visible || img.width<1 || img.height<1))
        throw new Error(`Missing or hidden image on page ${i+1}`);
      const file = path.join(folder, `page-${String(i+1).padStart(3,'0')}-${viewport.width}.png`);
      await page.screenshot({path:file,fullPage:true});
      evidence.push({page:i+1,rendered:true,viewport,file:await ref(file),layout,images});
    }
  }
  done('display', {pages:count, screenshots:evidence.length, widths:[1280,390]});
  await page.setViewportSize({width:1280,height:900});
  await page.locator('#pages').selectOption('0');
  await page.locator('#next').click();
  if (await page.locator('#pages').inputValue() !== '1') throw new Error('Next button failed');
  await page.locator('#prev').click();
  if (await page.locator('#pages').inputValue() !== '0') throw new Error('Previous button failed');
  await page.locator('main').click({position:{x:20,y:20}});
  await page.keyboard.press('ArrowRight');
  if (await page.locator('#pages').inputValue() !== '1') throw new Error('Keyboard failed');
  done('navigation', {next:true,previous:true,keyboard:true,selector_pages:count});
  await page.locator('#edit').check();
  const field = page.locator('.slide:visible [data-edit]').first();
  const original = await field.textContent();
  const edited = original + ' 浏览器编辑验证';
  await field.fill(edited);
  if (!(await page.locator('#state').textContent()).includes('草稿')) throw new Error('Missing draft warning');
  done('editing', {page:2,field:await field.getAttribute('data-edit'),draft_warning:true});
  const downloadWait = page.waitForEvent('download');
  await page.locator('#download').click();
  const download = await downloadWait;
  const exportFile = path.join(folder, 'exported-draft.html');
  await download.saveAs(exportFile);
  if (await download.failure()) throw new Error('Download failed');
  const fresh = await browser.newContext({offline:true, viewport:{width:1280,height:900}});
  const reopened = await fresh.newPage();
  watch(reopened);
  await reopened.goto(pathToFileURL(exportFile).href);
  await reopened.locator('#pages').selectOption('1');
  if (await reopened.locator('.slide:visible [data-edit]').first().textContent() !== edited)
    throw new Error('Export did not retain edits without cache');
  const restored = JSON.parse(await reopened.locator('[data-phase5-deck]').textContent());
  if (JSON.stringify(restored.deck_ref) !== JSON.stringify(request.target) || restored.draft.revision < 1)
    throw new Error('Export changed source reference');
  const reopenedImage = path.join(folder, 'export-reopened.png');
  await reopened.screenshot({path:reopenedImage,fullPage:true});
  done('export_reopen', {new_context:true,edit_restored:true,draft_revision:restored.draft.revision});
  if (errors.length || external.length) throw new Error('Page errors or external requests');
  done('offline', {original_context:true,reopen_context:true,external_requests:external.length});
  if ((await ref(source)).sha256 !== request.input_html.sha256) throw new Error('Original HTML changed');
  await fresh.close();
  await context.close();
  const result = {...request, mode:'deck', simulation:false, browser:'Chrome headless on Mac',
    environment, offline_context:true, checks, steps, evidence, errors, external,
    exported:await ref(exportFile), reopened_screenshot:await ref(reopenedImage)};
  await fs.writeFile(output, JSON.stringify(result,null,2) + '\n', {flag:'wx'});
} catch (error) {
  const reason = String(error.message).replaceAll(root,'<workspace>');
  await fs.writeFile(path.join(folder,'browser-failure.json'), JSON.stringify({
    status:'failed', checks, steps, errors, external, reason
  },null,2) + '\n', {flag:'wx'});
  console.error(reason);
  process.exitCode = 1;
} finally {
  if (browser) await browser.close();
}

File: scripts/deck_browser.py
"""Run the bundled browser probe before accepting its local evidence."""
import os
from pathlib import Path
import platform
import subprocess
import uuid

import phase5_common as c

PROTOCOL = "harness-deck-browser-v1"
RUNNER = "scripts/deck_browser.mjs"
CONTROLLER = "scripts/deck_browser.py"


def file_ref(root, relative):
    return {"path": relative, "sha256": c.sha(c.local(root, relative))}


def _bound_files(root, meta):
    html = next(item for item in meta["current_files"] if item["role"] == "html")
    return {"input_html": file_ref(root, html["path"]),
            "runner": file_ref(root, RUNNER), "controller": file_ref(root, CONTROLLER)}


def _text_output(value, root, limit=4000):
    if value is None:
        return ""
    if isinstance(value, bytes):
        value = value.decode("utf-8", errors="replace")
    text = str(value).replace(str(root), "<workspace>").strip()
    if len(text) <= limit:
        return text
    return text[:limit] + "\n[输出已截断]"


def _failure_kind(stdout, stderr, exit_code=None):
    text = f"{stdout}\n{stderr}".lower()
    if "module_not_found" in text or "cannot find module" in text:
        return "dependency_missing"
    if "unsupported browser" in text or ("protocol" in text and "unsupported" in text):
        return "configuration_error"
    if "timeout" in text or exit_code == -9:
        return "timeout"
    return "browser_process_failed"


def _write_failure(root, folder, *, stdout="", stderr="", exit_code=None,
                   reason="", error_type=""):
    clean_stdout = _text_output(stdout, root)
    clean_stderr = _text_output(stderr, root)
    kind = _failure_kind(clean_stdout, clean_stderr, exit_code)
    detail = _text_output(reason or clean_stderr or clean_stdout, root)
    c.atomic_json(folder / "failure.json", {
        "status": "failed",
        "failure_kind": kind,
        "exit_code": exit_code,
        "error_type": error_type or None,
        "reason": detail or kind,
        "stdout": clean_stdout,
        "stderr": clean_stderr,
    })
    return kind


def check_result(root, meta, result, run_id, bound):
    """Check the process outputs, not a claim supplied in a review report."""
    from phase5_visual import BROWSER_CHECKS, image_evidence, payload
    if not isinstance(result, dict) or any(result.get(k) != v for k, v in {
            "protocol": PROTOCOL, "run_id": run_id, "target": c.ref(meta),
            "mode": "deck", "simulation": False, "offline_context": True, **bound}.items()):
        raise c.WorkflowError("浏览器运行结果与本次执行或当前HTML不符")
    environment = result.get("environment", {})
    if (not isinstance(environment, dict) or environment.get("platform") != "darwin" or not environment.get("browser_version")
            or not environment.get("playwright_version")):
        raise c.WorkflowError("缺少实际Mac Chrome运行环境")
    if (result.get("checks") != dict.fromkeys(BROWSER_CHECKS, True)
            or result.get("errors") != [] or result.get("external") != []):
        raise c.WorkflowError("浏览器运行未完成全部检查")
    steps = result.get("steps")
    if (not isinstance(steps, list) or any(not isinstance(x, dict) for x in steps)
            or [x.get("name") for x in steps] != list(BROWSER_CHECKS)
            or any(x.get("passed") is not True or not x.get("observed") for x in steps)):
        raise c.WorkflowError("缺少实际交互步骤结果")
    expected = {(page, width) for page in range(1, len(payload(root, meta)["pages"]) + 1)
                for width in (1280, 390)}
    screenshots = result.get("evidence")
    if (not isinstance(screenshots, list) or len(screenshots) != len(expected)
            or any(not isinstance(item, dict) or not isinstance(item.get("viewport"), dict)
                   for item in screenshots) or {
            (item.get("page"), item.get("viewport", {}).get("width")) for item in screenshots} != expected):
        raise c.WorkflowError("浏览器运行未覆盖全部桌面及窄屏页面")
    folder = f"project/reviews/browser/{run_id}/"
    files = [result.get("exported"), result.get("reopened_screenshot"),
             *[item.get("file") for item in screenshots]]
    for item in files:
        if (not isinstance(item, dict) or not str(item.get("path", "")).startswith(folder)
                or not c.valid_event_file(root, item)):
            raise c.WorkflowError("浏览器运行副本或截图丢失、改变或不属于本次执行")
    for item in screenshots:
        if item.get("rendered") is not True:
            raise c.WorkflowError("存在未渲染页面")
        image_evidence(root, item["file"])
    image_evidence(root, result["reopened_screenshot"])
    from html_deck import _Metadata
    import json
    exported = c.local(root, result["exported"]["path"]).read_text(encoding="utf-8")
    data = json.loads("".join(_Metadata(exported).parts))
    if data.get("deck_ref") != c.ref(meta) or data.get("draft", {}).get("revision", 0) < 1:
        raise c.WorkflowError("导出副本未保留本次编辑及上游版本")


def run(root, key):
    from html_deck import _live
    if platform.system() != "Darwin":
        raise c.WorkflowError("浏览器验收目前仅支持Mac")
    root = Path(root).resolve()
    meta, _, _ = _live(root, key)
    bound = _bound_files(root, meta)
    # Only this fixed executable is dispatched; callers cannot register JSON as a run.
    for name, relative in (("runner", RUNNER), ("controller", CONTROLLER)):
        if bound[name]["sha256"] != c.sha(Path(__file__).with_name(Path(relative).name)):
            raise c.WorkflowError("工作包执行器与当前运行脚本不一致，请先对齐工作包")
    run_id = "deck-" + uuid.uuid4().hex
    folder = c.local(root, f"project/reviews/browser/{run_id}")
    folder.mkdir(parents=True, exist_ok=False)
    request = {"protocol": PROTOCOL, "run_id": run_id, "target": c.ref(meta), **bound}
    c.atomic_json(folder / "request.json", request)
    node = os.environ.get("HARNESS_NODE", "node")
    try:
        process = subprocess.run(
            [node, str(c.local(root, RUNNER)), str(root),
             (folder / "request.json").relative_to(root).as_posix()],
            capture_output=True, text=True, timeout=600, check=False)
    except (OSError, subprocess.SubprocessError) as exc:
        kind = _write_failure(root, folder, reason=str(exc) or type(exc).__name__,
                              error_type=type(exc).__name__)
        raise c.WorkflowError(f"浏览器未完成运行（{kind}）；已保留失败记录，没有登记成功") from exc
    if process.returncode:
        kind = _write_failure(root, folder, stdout=process.stdout, stderr=process.stderr,
                              exit_code=process.returncode)
        raise c.WorkflowError(f"浏览器检查失败（{kind}）；请查看本次运行目录，没有登记成功")
    try:
        evidence = file_ref(root, (folder / "browser-result.json").relative_to(root).as_posix())
        result = c.read_json(c.local(root, evidence["path"]))
        check_result(root, meta, result, run_id, bound)
        latest, _, _ = _live(root, key)
        if c.ref(latest) != c.ref(meta) or _bound_files(root, latest) != bound:
            raise c.WorkflowError("浏览器运行期间文件或上游已改变，须重新检查")
    except (OSError, ValueError, KeyError, TypeError, c.WorkflowError) as exc:
        kind = _write_failure(root, folder, reason=str(exc) or type(exc).__name__,
                              error_type=type(exc).__name__)
        raise c.WorkflowError(f"浏览器输出不完整或执行期间版本改变（{kind}），没有登记成功") from exc
    c.append_event(root, "browser_completed", {"protocol": PROTOCOL, "run_id": run_id,
        "target": c.ref(meta), "result": evidence, **bound})
    return {"run_id": run_id, "result": evidence, "pages": len(result["evidence"]) // 2,
            "checks": result["checks"]}


def verify_recorded_run(root, meta, result, evidence):
    runs = [event for event in c.events(root, "browser_completed")
            if event.get("run_id") == result.get("run_id") and event.get("result") == evidence]
    if len(runs) != 1:
        raise c.WorkflowError("缺少工作包实际执行记录；须先运行 deck-browser-check")
    event = runs[0]
    bound = _bound_files(root, meta)
    if (event.get("protocol") != PROTOCOL or event.get("target") != c.ref(meta)
            or any(event.get(k) != v for k, v in bound.items())):
        raise c.WorkflowError("浏览器执行记录对应的HTML、执行器或版本已改变")
    check_result(root, meta, result, event["run_id"], bound)
